### ISSUE TITLE
Type annotations "look through" typedefs, which can be confusing.

### ISSUE THREAD

**Issue by umanwizard** [rust-analyzer/rust-analyzer#1666]
blah.rs:
```rust
pub struct Foo;
```

lib.rs:
```rust
mod blah;
type BlahFoo = blah::Foo;
pub struct Foo {
    inner: BlahFoo,
}
pub fn f(f: Foo) {
    let _x = f.inner;
}
```

Rust-analyzer reports the type of `_x` to be `Foo`, which is quite confusing -- it would be better as `BlahFoo` or even `blah::Foo`

**Comment by matklad** [comment]
Yeah, I wonder what is the correct way to deal with type aliases....

Currently we eagerly normalize them, which causes the reported issue. I wonder if we should have `TypeAlias` as a variant of `TypeCtor::Adt`? That means though, that we'd need to adjust type equality, unification to properly deal with aliases.

That should make the following code possible ideally (which is currently an error in rustc):

```
type List<T> = (T, Box<Option<List<T>>>);
```

**Comment by flodiebold** [comment]
I personally don't think we should treat type aliases as a separate type. That seems like too much of a change from Rust's actual type system. (Also note that type aliases can alias to any type, not just ADTs.)

What I imagine is keeping track of where a type came from during type inference (either using a side table or using additional fields on the types), and then using that to represent the type in the way it was originally written (i.e. with type aliases, using the same path, leaving out default type parameters if they were left out there, etc.). Though I have to admit it's only a vague notion so far :smile:

**Comment by flodiebold** [comment]
(Or maybe put another way, it seems a layering violation to fix a 'display' issue like this by changing the type system.)

**Comment by bjorn3** [comment]
```rust
struct Unit;
type UnitAlias = Unit;

fn new_unit_alias() -> UnitAlias { Unit }

let a = new_unit_alias();
let b: Unit = a;
```

Should `a` be shown here as having type `Unit` or `UnitAlias`? When looking at the initializer first, you would see `UnitAlias`, while looking at the usage first, you would see `Unit`. And what if `Unit` and `UnitAlias` are swapped?

**Comment by flodiebold** [comment]
I think `a` should be shown as `UnitAlias` (but `b` as `Unit`). (Technically, when determining the type of `a`, the type annotation is only used to guide inference and unified with the actual inferred type in the end, so we would just have to take care of labeling the correct type as the 'expected' one, which is also important for diagnostics anyway.)

Maybe we would need a setting or an option to show 'normalized' types though.

**Comment by bjorn3** [comment]
@flodiebold :+1: for both.

**Comment by matklad** [comment]
@flodiebold yeah, doing this *for display* is centrally both wrong and insufficient (there are cases where you need to guess aliases back anyway).

However, fixing 

```
type List<T> = (T, Box<Option<List<T>>>);
```

seems like a good idea, if that's actually possible (don't know what it takes to unify `(T, Box<Option<List<T>>>) and `(T, Box<Option<List<(T, Box<Option<List<T>>>)>>>)`).

Also, this seems like a really good question for type-inference libraryfication, and I hope we'll start on than after nikomatsakis is back.

**Comment by umanwizard** [comment]
Actually I think there are two separate issues here, one about typedefs and one about paths. To state the second one more clearly, what if we just had this:

```
struct Foo {
    inner: blah::Foo,
}
```

then how should we display the type of `inner` ?

**Comment by umanwizard** [comment]
This is a bit of a tangent, but is anyone aware of what heuristics Clang uses in its error messages? C++ std lib types can be horrible (pages and pages of template crap), but Clang diagnostics always seems to choose a spelling that is reasonably easy to understand.

**Comment by bluss** [comment]
It would be so cool if type aliases could be preserved in the user displayed types, at least to some degree.

In the following example/screenshot, the function returns the type `Array3<f64>` which is a type alias for `ArrayBase<OwnedRepr<f64>, Ix3>` which is a type alias for `ArrayBase<OwnedRepr<f64>, Dim<[usize; 3]>>`. And I've heard of many crates that have worse types 😅 

Using type aliases would help users see through the too-complex types, and help them write better code too (use the type aliases).

![image](https://user-images.githubusercontent.com/3209739/80514317-9bbe9880-8980-11ea-8f62-f15d79e3c403.png)

**Comment by nbigaouette** [comment]
`uom` types are quite long. Say for example a [`Length`](https://docs.rs/uom/0.31.1/uom/si/length/type.Length.html) is a type alias, containing a type alias ([`Dimension`](https://docs.rs/uom/0.31.1/uom/si/length/type.Dimension.html)) containing other type alias...

If I don't limit the length of inlays types (`rust-analyzer.inlayHints.maxLength`), I get extremely long inlays of multiple time my screen size:

<img width="1267" alt="Capture d’écran, le 2021-03-26 à 20 26 02" src="https://user-images.githubusercontent.com/112988/112704624-a542d000-8e71-11eb-8bf0-07efd586adf2.png">

In that project I define type aliases for vectors of uom quantities. I would like to see those type aliases used as inlays and not the full length exact type.

Thanks!

**Comment by SomeoneToIgnore** [comment]
Another thread on this regard: https://rust-lang.zulipchat.com/#narrow/stream/185405-t-compiler.2Frust-analyzer/topic/Type.20aliases.20in.20inlay.20hints that makes me reconsider the labels we have on this one.

**Comment by flodiebold** [comment]
Unactionable doesn't mean anyone is prevented from working on it in any way, just that we don't see a direct way of making progress on it right now.

**Comment by weiznich** [comment]
To just give a example involving diesel where rust-analyzer shows private types as type hint:

```rust
#[macro_use] extern crate diesel; // 1.4.6
use diesel::prelude::*;

table! {
    users {
        id -> Integer,
        name -> Text,
    }
}

fn test() {
    let f = users::table.filter(users::name.eq("bar"));
}
```

which shows `SelectStatement<table, DefaultSelectClause, NoDistinctClause, WhereClause<{unknown}>>` for as type hint for `f`. The correct type would be `Filter<users::table, Eq<users::name, &'static str>>` as it is constructed by the [corresponding trait method](https://docs.diesel.rs/1.4.x/diesel/query_dsl/trait.QueryDsl.html#method.filter).

**Comment by iMplode-nZ** [comment]
Would it be possible to have like a setting `#[rust_analyzer::type_alias(X)]` on the structs for this purpose?

**Comment by HKalbasi** [comment]
What about generally trying to always using type alias for displaying types when they decrease complexity? We can define "complexity" as sum of generic arg counts, e.g. `Result<Box<dyn Trait<T>>, Error>` has complexity `2+1+1=4`, so `type Array3<T> = ArrayBase<OwnedRepr<T>, Ix3>` is a type alias that decreases complexity, but `type c_int = i32` is not. Then we can try to eagerly apply such type aliases when displaying types, even if the original expression doesn't use it.

This doesn't solve the problem completely, but is a step forward with relatively smaller change, and it is something that we may want to keep even when we support type aliases properly.

**Comment by Lokathor** [comment]
The type of a `c_int` should always show as `c_int`, or perhaps as `c_int = i32`.

Particularly with FFI code where the type's width can vary by platform, the stable API is often not "you can use i32" here, it's "you can use c_int here, whatever width that is"

**Comment by HKalbasi** [comment]
Yes, that wouldn't help in the `c_int` case. `c_int` needs proper support for lazy type aliases. (We definitely don't want to show all `i32`s in world as `c_int`).

**Comment by Lokathor** [comment]
ah, haha, yeah

**Comment by Ruhrpottpatriot** [comment]
I personally would like to have at least some notion of type aliases. I recently had `Result<(HashMap<String, HashMap<String, String>>, Option<Vec<String>>)>`  (don't ask why), which made my function signatures so horrible, that even clippy complained and suggested type aliases. Maybe using clippy's heuristic might be an avenue to tackle this?

Another thing I want to mention is the go extension for VSCode, which shows type aliases. I don't know to what extend, but for them it seems to work. Maybe their approach can be a solution?

**Comment by iago-lito** [comment]
In the hover text (since there is more room there than inlay hints), I would find it very useful to actually display the full aliasing stack:
```
TypeA
imported here with `use crate::mod1::A as TypeA;`
imported there with `use super::B as A;`
aliased there with `type B = external::LongTypeNameB;`
```

**Comment by dubrowgn** [comment]
I don't know if it's useful to point out rust analyzer currently appears inconsistent. Given:

```rust
type PacketIndex = SeqIndex;
```

Struct info shows base type (`SeqIndex`):

![Image](https://github.com/user-attachments/assets/fcc30b0b-b0c4-4e59-94c0-f5e65058efc9)

Function signature shows alias (`PacketIndex`, as expected):

![Image](https://github.com/user-attachments/assets/be1b85a5-0c9b-4c91-bd70-e35823fd3aaa)

Parameter info shows base type again:

![Image](https://github.com/user-attachments/assets/a2a52765-df9c-42a2-9722-12fb8a2cb88f)

Is there something special about the function signature that lets it work as expected that can't be used for the other cases?  (like, I dunno, it's just copy/pasting the raw text with no semantic understanding or something)

I can also do stuff like "Show all references to `PacketIndex`", and that works as expected (not showing me all instances of `SeqIndex`), so the appropriate info appears to be there and used sometimes, but not others.

**Comment by Veykril** [comment]
The difference is where the rendering of these derive their information from, the struct hover shows things based off of the non lowered forms (so we retain aliases as we never resolved them), where as function hovers actually re-render the fully resolved things. It is unrelatd to the concrete issue being asked about here though as that only works on the fully lowered type system information.

**Comment by Oblarg** [comment]
![Image](https://github.com/user-attachments/assets/6b6178a6-c7b8-45fa-95a2-77ad0df3ef66)

throwing my support behind this for my prototype units-of-measure library, which has a flat representation layer for the dimension data in the form of a *very long* const generic param list.  I have local aliases, and r-a ignores them, but the raw type is kind of useless to anyone who doesn't read library internals (and sucks even if you do).

This type should render as "Millimeters" or "Length" before it renders as `Quantity<...>`.  Ideally, it would show all three, ordered by locality-of-declaration and narrowness, like:

```
Millimeter;
Length;
Quantity<...>
```

EDIT: FWIW, i was able to (with AI assistance) write a LSP extension that intercepts the rust analyzer output and does quite a bit better:

![Image](https://github.com/user-attachments/assets/aac4e80b-bc37-4d40-b34f-9bdd63e80f7c)

so, for those struggling with this problem - you can probably work around it without too much trouble.  still should have first-class support though.

**Comment by clouds56** [comment]
Shall we add some experiment version of typedefs resolution first, and fix corner case after?

(Update: revise by copilot to be more fluent)
## Problem sketch
We have multiple typedefs (I'll call them "rules"), e.g.
```rust
pub type Array3<A> = Array<A, Ix3>;
pub type Array<A, D> = ArrayBase<OwnedRepr<A>, D>; // ArrayBase and OwnedRepr are structs (leafs)
```

Each typedef has a visibility / reachability (I'll call this its "scope"):
1. typedefs introduced in the current lexical scope
2. typedefs reachable from the current scope and marked pub
3. typedefs private in other modules

We also have source code that refers to Self/associated types, [for example](https://docs.rs/ndarray/latest/ndarray/type.Array.html#method.ones):
```rust
impl<A, D> Array<A, D>
where
    D: Dimension,
{
  pub fn ones<Sh>(shape: Sh) -> Self { ... }
}
```
Call the return type "text" (here Self).

## Solution
Suggested resolution algorithm (experimental)

1. Expand "text" locally without consulting typedef "rules":
    * Expand `Self` to `Array<A, D>`.
    * Substitute generics with the concrete types in the call site (e.g. `A = f64, D = Ix3`), producing `Array<f64, Ix3>`.
    * Recursively expand any further associated types / generics that are directly available in this local context.
    * Do NOT apply typedef rules in this phase (contrast with rustc, which would immediately apply `pub type Array<A, D> = ArrayBase<OwnedRepr<A>, D>;`).
2. After the local expansion, collect candidate typedef rules from scopes 1 and 2 (current lexical scope and reachable `pub` typedefs).
3. Attempt to apply those rules to the expanded type. In the example above:
    * `Array<f64, Ix3>` matches `Array<A, D>` with `A = f64, D = Ix3`.
    * Applying `pub type Array3<A> = Array<A, Ix3>;` yields `Array3<f64>`.
    * We can cache that `Array3<f64>` expands to `ArrayBase<OwnedRepr<f64>, Ix3>` for future use.

## Main difficulty: multiple applicable rules

When multiple rules apply, the question becomes how to combine them (and in which order) to produce the "most simplified" or most idiomatic type. This is similar to symbolic simplification (Mathematica's [`Simplify`](https://reference.wolfram.com/language/ref/Simplify.html)). A practical approach:

* Treat typedefs as rewrite rules and build a rewrite graph where nodes are type expressions and edges are rule applications.
* Search this graph for a normal form according to a cost metric, e.g. minimal textual length or minimal AST size. BFS or Dijkstra (cost = size) can find a minimal-cost representative.
* Ensure termination and avoid cycles by tracking visited nodes / applying a max-depth cutoff. Prefer local-scope rules when costs tie, or use heuristic tie-breakers such as prefer typedefs declared in current module.
* Cache results (mapping compact typedef instance -> fully expanded underlying representation) to avoid repeated work and to speed up future queries.

## Notes and trade-offs

* This experimental resolver is conservative compared to rustc's full expansion. It is intended for a better user-facing presentation in editor diagnostics, completion, and hover, not to replace rustc's type system.
* Be careful with private typedefs: only apply rules that are valid in the current context per visibility rules (scopes 1 and 2).
* Implement the minimal-form search incrementally: start with a deterministic heuristic (prefer direct typedef matches and current-scope typedefs), collect examples that disagree with rustc or are surprising, and evolve to a graph-search minimizer if needed.

**Comment by ChayimFriedman2** [comment]
@clouds56 The problem is not design; the problem is difficulty of implementation.
